SOLR-17968: HttpSolrClient as new base class

dsmiley opened the pull request:
https://issues.apache.org/jira/browse/SOLR-17968

The gist is:
 1. HttpApacheSolrClient no longer extends HttpSolrClient in this PR.  Unfortunate but it'll interfere...
 2. Rename HttpSolrClientBase to HttpSolrClient.  (replacing my earlier simple attempt at an abstract HttpSolrClient in this PR)
 3. Rename HttpSolrClientBuilderBase.java to HttpSolrClient.BuilderBase (inner class, now)
 4. Renamed HttpSolrClient.builder() internal instance method to `toBuilder(url)` to not conflict with...
 5. Added HttpSolrClient.builder(url) static method, mostly using code moved from CloudSolrClient, which now uses this.


----
Original outdated description:

Review the commits separately.  I'm likely to do a merge commit in the end.

Renamed {{org.apache.solr.client.solrj.apache.HttpSolrClient}} to {{HttpApacheSolrClient}}.  Note that HttpSolrClient *used* to be in {{org.apache.solr.client.solrj.impl}}.  Added {{org.apache.solr.client.solrj.HttpSolrClient}} as a pretty simple base class, which the obvious subclasses now as subclasses.  Notice its in the base SolrJ package (a package that we should carefully choose what goes there; shouldn't have many things).

dsmiley committed:
SOLR-17968: Rename HttpSolrClient to HttpApacheSolrClient

WILL NOT COMPILE

Not widely updating anything yet; just touch this source file.

dsmiley committed:
SOLR-17968: HttpSolrClient returns in org.apache.solr.client.solrj

WILL NOT COMPILE

Enhanced our clients to use it.

dsmiley committed:
ref guide sample code

dsmiley committed:
Reference HttpApacheSolrClient where appropriate.

dsmiley committed:
deprecations

dsmiley committed:
Don't override getBaseURL

dsmiley committed:
Change some HttpSolrClient references to HttpApacheSolrClient

dsmiley commented:
Come to think of it; I could split this between introducing `org.apache.solr.client.solrj.HttpSolrClient` (just touching a few files), and separately the rename of apache.HttpSolrClient to apache.HttpApacheSolrClient, which will include changing a number of import statements about where HttpSolrClient lives (touches many files).  No merge commit; just 2 separate commits that incrementally progress things.

dsmiley committed:
changelog

gerlowskija commented on solr/test-framework/src/java/org/apache/solr/client/solrj/apache/SolrPortAwareCookieSpecFactory.java:
```diff
@@ -31,6 +31,7 @@
 import org.apache.http.protocol.HttpContext;
 
 @SuppressWarnings("deprecation")
+@Deprecated
```
[Q] Why is this deprecated in this PR?  Is it just a part of the larger "apache client" deprecation that was missed until now?

dsmiley replied:
Yes; notice the package name.
This really underscores to me why our different implementations ought to get their very own packages!

gerlowskija reviewed (commented):
This LGTM by and large, but still have a few questions/concerns on the Javadocs/docs side.

Are there any ref-guide or other changes that should accompany this?  Upgrade notes maybe?  I saw the one change to `indexing-with-tika.adoc` page, but I imagine there'd be more needed?

gerlowskija commented:
FWIW - I raised a few questions about the general naming of these clients on [this mailing list thread](https://lists.apache.org/thread/84j5kc61n1ndt2ors1ozqw6dx93og30k).  I don't want that discussion to hold up this PR, but adding the link here as it is relevant if folks end up liking that idea...

dsmiley reviewed (commented):
From the list I said today:

> My only concern with bringing HttpSolrClient back (or adding SimpleSolrClient) is that it doesn't _really_ need to exist as a base abstraction.  In my PR, it just exposes a URL getter.  So what; the caller probably should have no need to call that any way.  It can be nice in a test that wants to manually then use an HttpClient, but it could just as well have gotten that URL from Jetty test infra.  


So I'm quite comfortable with simply not doing this issue/PR.  Well maybe some small tidbits of deprecating/moving a bit more but that's super minor.  Please let me know what you think.

I want to quickly move onto creating solrj-jetty, which I'd prefer not to do concurrently with this PR hanging around.

dsmiley committed:
Merge branch 'main' into SOLR-17968-HttpSolrClient

# Conflicts:
#	solr/core/src/test/org/apache/solr/cloud/HttpPartitionTest.java
#	solr/core/src/test/org/apache/solr/cloud/LeaderTragicEventTest.java
#	solr/core/src/test/org/apache/solr/cloud/LeaderVoteWaitTimeoutTest.java
#	solr/core/src/test/org/apache/solr/cloud/RecoveryZkTestWithAuth.java
#	solr/core/src/test/org/apache/solr/cloud/TestCloudConsistency.java
#	solr/core/src/test/org/apache/solr/cloud/TestRandomFlRTGCloud.java
#	solr/core/src/test/org/apache/solr/cloud/TestRandomRequestDistribution.java
#	solr/core/src/test/org/apache/solr/filestore/TestDistribFileStore.java
#	solr/core/src/test/org/apache/solr/handler/admin/CoreAdminHandlerTest.java
#	solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerFailureTest.java
#	solr/core/src/test/org/apache/solr/handler/admin/ZookeeperStatusHandlerTest.java
#	solr/core/src/test/org/apache/solr/handler/component/DistributedQueryElevationComponentTest.java
#	solr/core/src/test/org/apache/solr/pkg/TestPackages.java
#	solr/core/src/test/org/apache/solr/search/TestCoordinatorRole.java
#	solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpSolrClientBase.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/SolrExampleBinaryTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/SolrExampleCborTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/SolrExampleXMLTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/SolrExceptionTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/SolrSchemalessExampleTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/TestSolrJErrorHandling.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/embedded/SolrExampleJettyTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/response/InputStreamResponseParserTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/response/TestSuggesterResponse.java
#	solr/test-framework/src/java/org/apache/solr/SolrTestCaseJ4.java
#	solr/test-framework/src/java/org/apache/solr/client/solrj/apache/ConcurrentUpdateSolrClient.java
#	solr/test-framework/src/java/org/apache/solr/client/solrj/apache/HttpApacheSolrClient.java
#	solr/test-framework/src/java/org/apache/solr/client/solrj/apache/SolrClientBuilder.java
#	solr/test-framework/src/java/org/apache/solr/cloud/AbstractBasicDistributedZkTestBase.java
#	solr/test-framework/src/java/org/apache/solr/cloud/AbstractFullDistribZkTestBase.java
#	solr/test-framework/src/java/org/apache/solr/cloud/AbstractRecoveryZkTestBase.java
#	solr/test-framework/src/java/org/apache/solr/cloud/AbstractSyncSliceTestBase.java
#	solr/test-framework/src/java/org/apache/solr/cloud/AbstractUnloadDistributedZkTestBase.java
#	solr/test-framework/src/java/org/apache/solr/embedded/JettySolrRunner.java
#	solr/test-framework/src/test/org/apache/solr/client/solrj/apache/BasicHttpSolrClientTest.java
#	solr/test-framework/src/test/org/apache/solr/client/solrj/apache/HttpSolrClientBuilderTest.java

dsmiley committed:
HttpApacheSolrClient no longer implements HttpClient

dsmiley committed:
Rename HttpSolrClientBase to HttpSolrClient.
Rename HttpSolrClientBuilderBase.java to HttpSolrClient.BuilderBase (inner class)

dsmiley committed:
Add HttpSolrClient.builder(baseUrl) static method

dsmiley committed:
Merge branch 'main' into SOLR-17968-HttpSolrClient

# Conflicts:
#	solr/solr-ref-guide/modules/indexing-guide/pages/indexing-with-tika.adoc

dsmiley commented:
I did a series of things, as seen in a series of commits:
1. HttpApacheSolrClient no longer extends HttpSolrClient in this PR.  Unfortunate but it'll interfere...
2. Rename HttpSolrClientBase to HttpSolrClient.  (replacing my earlier simply attempt at an abstract HttpSolrClient in this PR)
3. Rename HttpSolrClientBuilderBase.java to HttpSolrClient.BuilderBase (inner class, now)
4. Renamed HttpSolrClient.builder() internal instance method to `toBuilder(url)` to not conflict with...
5. Added HttpSolrClient.builder(url) static method, mostly using code moved from CloudSolrClient, which now uses this.

dsmiley committed:
fix from merge

dsmiley committed:
Merge remote-tracking branch 'apache/main' into SOLR-17968-HttpSolrClient

# Conflicts:
#	solr/core/src/test/org/apache/solr/handler/TestReplicationHandlerBackup.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/TestBatchUpdate.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/TestSolrJErrorHandling.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/embedded/SolrExampleJettyTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/impl/HttpSolrClientBadInputTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/impl/HttpSolrClientTestBase.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/response/InputStreamResponseParserTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/response/TestSuggesterResponse.java
#	solr/test-framework/src/java/org/apache/solr/SolrJettyTestBase.java
#	solr/test-framework/src/test/org/apache/solr/client/solrj/apache/BasicHttpSolrClientTest.java

jdyer1 reviewed (approved):
This is very nice and really appreciate your being willing to do all this grunt work.

igiguere commented on solr/core/src/test/org/apache/solr/cloud/BasicDistributedZkTest.java:
```diff
@@ -699,7 +699,7 @@ public void test() throws Exception {
       assertEquals(
           "unexpected pre-commitWithin document count on node: "
-              + ((HttpSolrClient) client).getBaseURL(),
+              + ((HttpApacheSolrClient) client).getBaseURL(),
```
Here and elsewhere:
HttpApacheSolrClient is supposed to die, but all these tests now use it instead of HttpSolrClient: the new one should be used.
IMHO, this looks weird, and it will all have to change eventually anyways when HttpApacheSolrClient is removed.

dsmiley replied:
These tests continue to use the legacy client, despite its name changing.  There's a lot of work in updating the tests... I started it and shelved it for higher priorities: https://github.com/dsmiley/solr/pull/7   Pugh told me he's interested in taking over

igiguere replied:
Good for me, then.

igiguere replied:
@epugh : I'm on https://issues.apache.org/jira/browse/SOLR-17436 for now.  Maybe later...

dsmiley committed:
solrj-streaming need not depend on jetty

gerlowskija commented on solr/core/src/test/org/apache/solr/cloud/BasicDistributedZkTest.java:
```diff
@@ -1150,7 +1150,7 @@ protected CollectionAdminResponse createCollection(
     CollectionAdminResponse res = new CollectionAdminResponse();
     if (client == null) {
-      final String baseUrl = ((HttpSolrClient) clients.get(clientIndex)).getBaseURL();
+      final String baseUrl = ((HttpApacheSolrClient) clients.get(clientIndex)).getBaseURL();
```
[0] These casts to the specific type (`HttpApacheSolrClient`) are kindof unfortunate, given that `getBaseURL` exists on the more general `HttpSolrClient`.

You mentioned in your PR level comment that having `HttpApacheSolrClient` extend `HttpSolrClient` "interferes" in other places.  Can you give a bit more context on that decision here please?

dsmiley replied:
I'm about to push a revert of that rename in this PR (HttpApacheSolrClient goes back to HttpSolrClient).  Thus 2 such classes will exist in separate packages.  I have separate work that removes use of apache HttpSolrClient completely, and that which is nearly ready to share but I want this one in first.

gerlowskija commented on solr/test-framework/src/test/org/apache/solr/client/solrj/apache/BasicHttpSolrClientTest.java:
Your comment above suggests you intended to rename this.

I don't care strongly whether you do or not, just flagging this in case it got inadvertently lost in the shuffle.

gerlowskija reviewed (commented):
This looks fine overall, but I'm still a little struck by the lack of docs.  (Javadocs look great - referring more to higher-level docs like changelog entries, Upgrade Notes, etc.)

Surely tons of users were using HttpSolrClient despite the deprecation and could use an extra bit of help when they go to upgrade?  Surely lots of users will have used the builders whose names are changing here.

Am I missing something? (Say, that docs were put in back when the deprecations occurred.)

I'd happily "Approve" as soon as there's a bit more in that direction.

dsmiley committed:
Merge branch 'refs/heads/main' into SOLR-17968-HttpSolrClient

# Conflicts:
#	solr/core/src/test/org/apache/solr/cloud/TestTlogReplica.java
#	solr/core/src/test/org/apache/solr/filestore/TestDistribFileStore.java
#	solr/core/src/test/org/apache/solr/handler/admin/api/ClusterPropsAPITest.java
#	solr/core/src/test/org/apache/solr/metrics/SolrMetricsIntegrationTest.java
#	solr/core/src/test/org/apache/solr/pkg/TestPackages.java
#	solr/core/src/test/org/apache/solr/search/TestCoordinatorRole.java
#	solr/core/src/test/org/apache/solr/servlet/CacheHeaderTestBase.java
#	solr/core/src/test/org/apache/solr/servlet/SecurityHeadersTest.java
#	solr/solrj-jetty/src/java/org/apache/solr/client/solrj/jetty/HttpJettySolrClient.java
#	solr/solrj-streaming/gradle.lockfile
#	solr/solrj-streaming/src/java/org/apache/solr/client/solrj/io/SolrClientCache.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/SolrExampleBinaryTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/SolrExampleCborTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/SolrExampleXMLTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/SolrExceptionTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/SolrSchemalessExampleTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/TestLBHttpSolrClient.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/embedded/SolrExampleJettyTest.java
#	solr/solrj/src/test/org/apache/solr/client/solrj/impl/HttpSolrClientTestBase.java
#	solr/test-framework/src/java/org/apache/solr/SolrTestCaseJ4.java
#	solr/test-framework/src/java/org/apache/solr/cloud/SolrCloudTestCase.java
#	solr/test-framework/src/java/org/apache/solr/embedded/JettySolrRunner.java
#	solr/test-framework/src/java/org/apache/solr/util/RestTestBase.java
#	solr/test-framework/src/test/org/apache/solr/client/solrj/apache/BasicHttpSolrClientTest.java
#	solr/test-framework/src/test/org/apache/solr/client/solrj/apache/ConnectionReuseTest.java
#	solr/test-framework/src/test/org/apache/solr/client/solrj/apache/HttpSolrClientBuilderTest.java
#	solr/test-framework/src/test/org/apache/solr/client/solrj/apache/HttpSolrClientConPoolTest.java
#	solr/test-framework/src/test/org/apache/solr/client/solrj/apache/LBHttpSolrClientTest.java

dsmiley committed:
Restore HttpApacheSolrClient back to HttpSolrClient for now. Rollback test changes generally.

dsmiley commented on solr/test-framework/src/java/org/apache/solr/client/solrj/apache/package-info.java:
```diff
@@ -16,4 +16,5 @@
  */
 
 /** Apache HttpClient based {@link org.apache.solr.client.solrj.SolrClient} implementations */
+@Deprecated
```
CHANGELOG.md 10.0:
> Solr no longer uses Apache HttpClient. Nonetheless SolrJ still includes deprecated SolrClient implementations for that (which will be removed eventually), and many tests still use it. Also, there were server metrics relating to HTTP client connection pools that no longer exist. [SOLR-17872](https://issues.apache.org/jira/browse/SOLR-17872) (David Smiley)

dsmiley reviewed (commented):
Did a force merge unfortunately.  Reverted most test changes, shrinking this PR substantially.

TODO:  major-changes-in-solr-10.adoc

dsmiley committed:
Merge branch 'main' into SOLR-17968-HttpSolrClient

# Conflicts:
#	solr/solrj/src/test/org/apache/solr/client/solrj/impl/ClusterStateProviderTest.java

dsmiley commented on solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java:
```diff
@@ -21,8 +21,8 @@
 /**
- * This {@link CloudSolrClient} is a base implementation using a {@link HttpSolrClientBase}. The '2'
- * in the name has no differentiating significance anymore.
+ * This {@link CloudSolrClient} is a base implementation using a {@link HttpSolrClient}. The '2' in
```
the 2 refers to CloudHttp2SolrClient

dsmiley commented on solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpSolrClient.java:
```diff
@@ -104,7 +97,24 @@ protected HttpSolrClientBase(String serverBaseUrl, HttpSolrClientBuilderBase<?,
+  }
+
+  /** Typically looks like {@code http://localhost:8983/solr} (no core or collection) */
+  public String getBaseURL() {
```
I kinda wish we use URL as a type instead of a String both for added type safety and also likely avoids extra parsing.  Jetty & JDK HttpClients natively works with a URI -- passing a String results in URI parsing _anyway_.
I suppose we could add getBaseUri at some point.

dsmiley committed:
documentation

dsmiley reviewed (commented):
This is ready; I believe I addressed concerns.  I added more ref-guide documentation.

I plan to merge this Friday afternoon if I get no further feedback / requests to defer.

dsmiley committed:
Merge branch 'main' into SOLR-17968-HttpSolrClient

# Conflicts:
#	solr/core/gradle.lockfile
